query_optimizer/rules/UnnestSubqueries.cpp

```diff
@@ -0,0 +1,720 @@
+ *   limitations under the License.
+ **/
+
+#include "query_optimizer/rules/UnnestSubqueries.hpp"
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "query_optimizer/OptimizerContext.hpp"
+#include "query_optimizer/expressions/AttributeReference.hpp"
+#include "query_optimizer/expressions/ComparisonExpression.hpp"
+#include "query_optimizer/expressions/Exists.hpp"
+#include "query_optimizer/expressions/ExprId.hpp"
+#include "query_optimizer/expressions/ExpressionType.hpp"
+#include "query_optimizer/expressions/ExpressionUtil.hpp"
+#include "query_optimizer/expressions/LogicalAnd.hpp"
+#include "query_optimizer/expressions/LogicalNot.hpp"
+#include "query_optimizer/expressions/LogicalOr.hpp"
+#include "query_optimizer/expressions/NamedExpression.hpp"
+#include "query_optimizer/expressions/PatternMatcher.hpp"
+#include "query_optimizer/expressions/SubqueryExpression.hpp"
+#include "query_optimizer/logical/Aggregate.hpp"
+#include "query_optimizer/logical/Filter.hpp"
+#include "query_optimizer/logical/HashJoin.hpp"
+#include "query_optimizer/logical/Logical.hpp"
+#include "query_optimizer/logical/LogicalType.hpp"
+#include "query_optimizer/logical/PatternMatcher.hpp"
+#include "query_optimizer/logical/Project.hpp"
+#include "query_optimizer/logical/TopLevelPlan.hpp"
+#include "query_optimizer/rules/Rule.hpp"
+#include "types/operations/comparisons/Comparison.hpp"
+#include "types/operations/comparisons/ComparisonID.hpp"
+#include "utility/SqlError.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+namespace optimizer {
+
+namespace E = ::quickstep::optimizer::expressions;
+namespace L = ::quickstep::optimizer::logical;
+
+struct CorrelatedQueryInfo {
+  enum QueryType {
+    kHashJoin,
+    kSemiJoin,
+    kAntiJoin
+  };
+
+  CorrelatedQueryInfo(const QueryType query_type,
+                      const L::LogicalPtr &correlated_query_in,
+                      std::vector<E::AttributeReferencePtr> &&probe_join_attributes_in,
+                      std::vector<E::AttributeReferencePtr> &&build_join_attributes_in,
+                      std::vector<E::PredicatePtr> &&non_hash_join_predicates_in)
+      : query_type_(query_type),
+        correlated_query(correlated_query_in),
+        probe_join_attributes(std::move(probe_join_attributes_in)),
+        build_join_attributes(std::move(build_join_attributes_in)),
+        non_hash_join_predicates(std::move(non_hash_join_predicates_in)) {}
+
+  QueryType query_type_;
+  L::LogicalPtr correlated_query;
+  std::vector<E::AttributeReferencePtr> probe_join_attributes;
+  std::vector<E::AttributeReferencePtr> build_join_attributes;
+  std::vector<E::PredicatePtr> non_hash_join_predicates;
+};
+
+L::LogicalPtr UnnestSubqueries::apply(const L::LogicalPtr &input) {
+  DCHECK(L::SomeTopLevelPlan::Matches(input))
+      << "The input logical node must be of TopLevelPlan";
+
+  const L::TopLevelPlan* top_level_plan =
```

Comment:
What do you think of applying the pointer sigil style here?

`const L::TopLevelPlan *top_level_plan`.